FutureFeatures.md

```diff
@@ -66,16 +66,160 @@ Options under consideration:
+   values passed to a WebIDL invocation to the signature declared by (1) and
+   checked by (2).
+
+(1) and (2) of the WebIDL spec are meaningful to WebAssembly, but (3)
+would effectively be skipped.
+
+Another important issue is mapping WebIDL values types that aren't simple
+[primitive types](http://www.w3.org/TR/WebIDL/#dfn-primitive-type):
+* [Dictionary types](http://www.w3.org/TR/WebIDL/#idl-dictionary)
+  would [appear](http://www.w3.org/TR/WebIDL/#es-dictionary) to require
+  JS objects but are actually defined as values such that they can
+  be (and are, in various browser implementations) flattened to C structs.
+  Thus, a natural WebAssembly binding would be to map dictionaries to structs
+  in linear memory passed by reference (integer offset).
+* The same goes for [sequence types](http://www.w3.org/TR/WebIDL/#idl-sequence).
+* [Enumeration types](http://www.w3.org/TR/WebIDL/#es-enumeration) could be
+  mapped to canonical integers.
+* [Union types](http://www.w3.org/TR/WebIDL/#idl-union) could be handled in
+  multiple ways. One option is to treat the union type itself as an importable
+  opaque reference type (when all the elements are themselves reference types).
+  Another option is to introduce an overload of each signature for each element
+  of the union type such that all calls passed a single element type and the
+  full Union Type was never explicitly represented in WebAssembly.
+* [Callback function types](http://www.w3.org/TR/WebIDL/#es-callback-function)
+  could map to a `(function pointer, environment pointer)` closure pair.
+
+Overall, the goal of mapping WebIDL to WebAssembly builtin modules is to avoid
+the need to define a duplicate WebAssembly interface for all Web APIs.  In
+practice, some WebIDL patterns may have an unnatural or inefficient mapping
+into WebAssembly such that new overloads and best practices would need to be
+adopted. Over time, though, these rough edges would be ironed out leaving the
+long term benefit of defining Web APIs with a single interface and ensuring
+that JS and WebAssembly always had access to the same raw functionality.
+
+### Direct GC access
+
+In contract to *opaque* reference types, a second sub-feature would be to allow
+direct GC allocation and field access from WebAssembly code through
+*non-opaque* reference types.
+
+There is a lot of the design left to
+consider for this feature, but a few points of tentative agreement are:
+* To avoid baking in a single language's object model, define low-level GC
+  primitives (viz., structs and arrays) and allow the source language compiler
+  to build up features like virtual dispatch and access control.
+* GC struct and array types would have associated *struct/array reference
+  types* that were similar to and symmetric with 
+  [opaque reference types](FutureFeatures.md#opaque-reference-types)
+  (just not opaque).
+* The GC heap would be semantically distinct from linear memory and thus
```

Comment:
IIRC .NET maintains a single native stack, which is safe-enough because the only way to get the address of something on the stack (or read/write directly from addresses) is to drop into ```unsafe``` mode. Anyone using pointers in C# is interacting with this too, even though C++/CLI's access is more powerful.

.NET structs allocated on the GC heap either live in a box or inside a reference type, so they're never actually allocated *directly* on the GC heap. They are effectively native values, just like in C. They do have the ability to hold object references, though, so in my experience the runtime rejects any attempt to store any struct containing GC references into the native heap - only allowed to put it on the stack or in the GC heap.

I think their model works pretty well, but we have more stringent safety/robustness requirements, and we can't protect the stack as easily as they can since raw memory access is a baseline for us.